### PR TITLE
document unittest.nim + code agreement

### LINKED ISSUE #460: Better docs for unittest module
The unittest module has very few docs and the only way currently to get started with it is to read the examples in the git repo which is not exactly ideal.

### PR DESCRIPTION
**Pull request by fenekku**
Couple of things:

- document unittest.nim and thus fix #460
- make `require` agree with prior stated doc/intent: if `require` fails we directly exit, do not pass go and do not collect 200$.
- `expect` will now fail gracefully if an exception is not thrown AND/OR if an unexpected exception is raised.
- first patch in a set of at least another patch

I tried to follow some of the documentation guides:

- https://github.com/nim-lang/Nim/blob/devel/contributing.rst
- https://github.com/nim-lang/Nim/issues/3089

Please review! I would love it if *all* the Nim documentation was consistent
and felt as though it was written by the same person.

Discussion points/next patch:

- `testSetupIMPL` and `testTeardownIMPL` are public right now, but I don't think they
  should be. We should deprecate their publicness. That is why I didn't document them.
- I want `check` to actually print out the values that cause the false expression when those values are not literals. I am working on that. Only one level-deep of `repr` for these values is fine. Should `repr` even be used for that?
- I have added a test, but how do you test tests? Especially failure cases? I would be interested in hearing ideas about that!

### CODE REVIEW

File: lib/pure/unittest.nim [hunk 1]
```diff
@@ -143,7 +235,7 @@ macro check*(conditions: stmt): stmt {.immediate.} =
       checkpoint(name & " was " & $value)
 
   proc inspectArgs(exp: NimNode) =
-    for i in 1 .. <exp.len:
+    for i in countup(1, exp.len - 1):
```
Comment (dom96) [inline lib/pure/unittest.nim]: Why did you change this?

**Comment by dom96** [comment]
I'm going to merge this because it looks good, and this PR has waited long enough (sorry!).

**Comment by dom96** [comment]
> testSetupIMPL and testTeardownIMPL are public right now, but I don't think they should be. We should deprecate their publicness. That is why I didn't document them.

We can likely remove those if they are not used at all. Maybe @zah can chime in.

> I want check to actually print out the values that cause the false expression when those values are not literals. I am working on that. Only one level-deep of repr for these values is fine. Should repr even be used for that?

Maybe `$` would work too?

> I have added a test, but how do you test tests? Especially failure cases? I would be interested in hearing ideas about that!

`when isMainModule` + `assert`? i.e. the traditional way. I'm not sure really.

File: lib/pure/unittest.nim (hunk 1 above)
Comment (fenekku) [inline lib/pure/unittest.nim]: Consistency with the previous style :) .

**Comment by fenekku** [comment]
Given the issue #2857 I had with invisible terminal codes,  I thought `repr` might be good. I am not a big fan of `repr`'s readability so I would be happy with `$`.

> I'm going to merge this because it looks good, and this PR has waited long enough (sorry!).

I am actually 3/4 of the way for the next patch so I am the one who is slow :blush: .
